src/core/plugins.js

```diff
@@ -0,0 +1,66 @@
+(function() {
+    'use strict';
+
+    /**
+     * Wraps each of the plugin lifecycle methods in a closure that will
+     * set up the editor.__processingPlugin__ variable so it can be globally
+     * accessed exposing the plugin being processed and the lifecycle phase
+     * in which it is happening
+     *
+     * @private
+     * @method wrapPluginLifecycle
+     * @param {Object} plugin The plugin to wrap lifecycle methods
+     */
+    var wrapPluginLifecycle = function(plugin) {
+        var methods = ['beforeInit', 'init', 'afterInit'];
+
+        methods.forEach(function(methodName) {
+            if (plugin[methodName]) {
+                plugin[methodName] = CKEDITOR.tools.override(plugin[methodName], function(originalPluginMethod) {
+                    var payload = {
+                        phase: methodName,
+                        plugin: plugin
+                    };
+
+                    return function(editor) {
+                        editor.__processingPlugin__ = payload;
+
+                        originalPluginMethod.call(this, editor);
+
+                        delete editor.__processingPlugin__;
```

Comment:
`delete` is better to be still avoided, haven't heard engine developers resolved the performance penalties of using it. Would it be possible to use `editor.__processingPlugin__ = null` here?